### PR TITLE
[RFC] Support functions in custom clipboard providers

### PR DESCRIPTION
**Pull request by equalsraf**
Custom clipboard providers can be set using g:clipboard, where
g:clipboard['copy']['+'] was originally meant to be a string with
a command that sets the clipboard.

This commit enables the use of vimscript functions instead. The function
signatures is the same as the clipboard provider itself. A clipboard
provider is expected to store and return a list of lines (i.e. the text)
and a register type (as seen in setreg()).

The cache_enabled attributed is completely ignored if the 'copy'
attribute is set to be a function.

## Why

Primarily to enable integration with clipboards provided by an external UI. But I'm sure someone will find other uses for this.

Originally I tried doing this by overriding the provider in https://github.com/equalsraf/neovim-qt/pull/436 it worked partially (originally from [neovim-gtk](https://github.com/daa84/neovim-gtk/blob/master/runtime/plugin/nvim_gui_shim.vim#L7)) but you run into problems when you want to attach and reattach the UI, because you cannot override the provider again.

##  Reasons not to do this

- we should not expose the internals of the clipboard provider
- the clipboard provider lines go through systemlist() which converts NUL into NL (i think) I dont think this is documented as part of the clipboard, but custom providers would have to conform to this too.
- other approaches e.g. does this overlap with #4448?

### CODE REVIEW

**Comment by bfredl** [comment]
This would be useful also for https://github.com/bfredl/nvim-x11. Arguably this makes external implementors _less_ coupled to nvim internals than if they were overriding `autoload/provider` directly. It is also simpler to do something like

    if has_xorg_server
      let g:clipboard = nvim_x11_clipboard
    endif

> NUL into NL (i think) I dont think this is documented as part of the clipboard, but custom providers would have to conform to this too.

This is the canonical representation of arbitrary binary data in vimL. It is used by getreg/setreg and functions that get or set buffer lines.

File: runtime/autoload/provider/clipboard.vim [hunk 1]
```diff
@@ -127,10 +138,14 @@ if empty(provider#clipboard#Executable())
 endif
 
 function! s:clipboard.get(reg) abort
-  if s:selections[a:reg].owner > 0
-    return s:selections[a:reg].data
+  if type(s:paste[a:reg]) == v:t_func
+    return s:paste[a:reg]()
+  else
+    if s:selections[a:reg].owner > 0
```
Comment (justinmk) [inline runtime/autoload/provider/clipboard.vim]: Could use `elseif` to avoid nesting

Comment (equalsraf) [reply runtime/autoload/provider/clipboard.vim]: Moved this into an elseif

**Comment by equalsraf** [comment]
I've added some tests and docs with regards to the paste function, as it turns out it can return a number of types.

File: runtime/doc/provider.txt [hunk 2]
```diff
@@ -178,6 +178,29 @@ If `cache_enabled` is |TRUE| then when a selection is copied, Nvim will cache
...
+          \      '*': {lines, regtype -> extend(g:, {'dummy_clipboard_star': [lines, regtype]}) },
+          \    },
+          \   'paste': {
+          \      '+': {-> get(g:, 'dummy_clipboard_plus', [])},
+          \      '*': {-> get(g:, 'dummy_clipboard_star', [])},
+          \   },
+          \ }
+
+The copy function is meant to place a list of lines into the clipboard and
+store a register type specification along with the data.
```
Comment (bfredl) [inline runtime/doc/provider.txt]: It could be worth mentioning the register types: `v` characterwise, `V` linewise and `b` blockwise.

Comment (justinmk) [reply runtime/doc/provider.txt]: I added a  link to `|setreg()|` when merging.

Comment (equalsraf) [reply runtime/doc/provider.txt]: ooops I just push a change for this

**Comment by justinmk** [comment]
Merged after fixing indent and adjusting the docs. Thanks @equalsraf

**Pull request closed without merging**